### DISCUSSION TITLE
Configuring a staging environment in Shakapacker 7

### DISCUSSION THREAD

**Discussion by jpwilksch** [shakacode/shakapacker#337] [Q&A] [answered]
I made my way through the upgrade docs, upgrading from Webpacker 5 through to Shakapacker 7. Quite a few bumps along the way, but I got there.

Now after completing the migration, deployment to our staging environment is broken because the Shakapacker compile fails.

Currently I have the development, test and production environments defined in config/shakapacker.yml. When I manually run the asset precompile task, Shakapacker correctly fallsback to `production` environment because `staging` is not defined there, and the compile completes successfully (the 2 warnings are for asset size). All good.

```
$ RAILS_ENV=staging bundle exec rake assets:precompile
RAILS_ENV=staging environment is not defined in config/shakapacker.yml, falling back to production environment
...
I, [2023-07-14T13:32:51.647630 #107062]  INFO -- : Compiling...
I, [2023-07-14T13:33:53.284271 #107062]  INFO -- : Compiled all packs in /home/user/portal/public/packs
E, [2023-07-14T13:33:53.284383 #107062] ERROR -- : Warning: 'staging' environment not found in the configuration. Using 'production' configuration as a fallback.
...
webpack 5.88.1 compiled with 2 warnings in 59070 ms
```

When I deploy via Capistrano, I again see Shakapacker falling back to `production` environment:
```
00:11 deploy:assets:precompile
      01 bundle exec rake assets:precompile
      01 RAILS_ENV=staging environment is not defined in config/shakapacker.yml, falling back to production environment
...
```

However, the following error is then encountered and the deployment fails:
```
...
COMPILATION FAILED:
EXIT STATUS: pid 44597 exit 1
OUTPUTS:
assets by status 6.09 KiB [cached] 3 assets
Entrypoint application = js/runtime-179962414133248460dd.js js/application-d4f033fdacda43b96dc8.js 1 auxiliary asset
runtime modules 2.62 KiB 3 modules
./app/javascript/packs/application.js 39 bytes [built] [code generated] [1 error]

WARNING in DefinePlugin
Conflicting values for 'process.env.NODE_ENV'

1 warning has detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

ERROR in ./app/javascript/packs/application.js
Module build failed (from ./node_modules/babel-loader/lib/index.js):
Error: [BABEL] /var/www/rails/portal/releases/20230714034636/app/javascript/packs/application.js: Please specify a valid NODE_ENV or BABEL_ENV environment variable. Valid values are "development", "test", and "production". Instead, received: ""staging"". (While processing: "/var/www/rails/portal/releases/20230714034636/node_modules/shakapacker/package/babel/preset.js")
    at config (/var/www/rails/portal/releases/20230714034636/node_modules/shakapacker/package/babel/preset.js:11:11)
    at async (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/gensync-utils/async.js:36:33)
    at async (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:186:15)
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:216:13
    at Generator.next (<anonymous>)
    at /var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/full.js:181:21
    at Generator.next (<anonymous>)
    at Function.<anonymous> (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/gensync-utils/async.js:21:3)
    at Generator.next (<anonymous>)
    at step (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:269:25)
    at evaluateAsync (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:291:5)
    at Function.errback (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:113:7)
    at errback (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/gensync-utils/async.js:66:18)
    at async (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:188:17)
    at onFirstPause (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:216:13)
    at Generator.next (<anonymous>)
    at cachedFunction (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/caching.js:52:46)
    at cachedFunction.next (<anonymous>)
    at loadPresetDescriptor (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/full.js:293:43)
    at loadPresetDescriptor.next (<anonymous>)
    at recursePresetDescriptors (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/full.js:77:31)
    at recursePresetDescriptors.next (<anonymous>)
    at /var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/full.js:157:21
    at Generator.next (<anonymous>)
    at loadFullConfig (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/config/full.js:113:5)
    at loadFullConfig.next (<anonymous>)
    at transform (/var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/transform.js:20:45)
    at transform.next (<anonymous>)
    at step (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:269:25)
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:273:13
    at async.call.result.err.err (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:223:11)
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:189:28
    at /var/www/rails/portal/releases/20230714034636/node_modules/@babel/core/lib/gensync-utils/async.js:68:7
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:113:33
    at step (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:287:14)
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:273:13
    at async.call.result.err.err (/var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:223:11)
    at /var/www/rails/portal/releases/20230714034636/node_modules/gensync/index.js:37:40
```

The error originates from `./node_modules/shakapacker/package/babel/preset.js` and the cause appears to be that it is being passed "staging" where it should be "production":

> Please specify a valid NODE_ENV or BABEL_ENV environment variable. Valid values are "development", "test", and "production". Instead, received: "**"staging"**".

For staging environment on Webpacker 5 we had to manually set `process.env.NODE_ENV` to production and again with `DefinePlugin`, but from the [Shakapacker 7 docs](https://github.com/shakacode/shakapacker#custom-rails-environments) it would appear we no longer have to do this. 

The same error occurs when I add a copy of the production config as "staging" to config/shakapacker.yml.

We have a nearly-default `webpack.config.js`:
```
const { generateWebpackConfig, merge } = require('shakapacker')
const webpack = require('webpack')

const webpackConfig = generateWebpackConfig()
const customConfig = {
  devServer: {
    watchFiles: {
      options: {
        ignored: '**/.*'
      }
    }
  },
  plugins: [
    new webpack.ProvidePlugin({
      $: 'jquery',
      jQuery: 'jquery',
    })
  ]
}

module.exports = merge(webpackConfig, customConfig)
```

What have I overlooked?

**Comment by jpwilksch** [comment]
I found a solution by re-instating the `process.env.NODE_ENV` logic we previously had in our config/webpack/staging.js configuration for Webpack 5. 

The [Custom Rails Environments](https://github.com/shakacode/shakapacker#custom-rails-environments) section of the docs does not mention this as a requirement, and almost implies that Shakapacker takes care of this conversion from RAILS_ENV to correct NODE_ENV for you. If my interpretation _is_ correct, perhaps I have some other configuration error that prevents this working.

> Otherwise, Shakapacker will use the production environment as a fallback environment for loading configurations. Please note, NODE_ENV can either be set to production, development or test. This means you don't need to create additional environment files inside config/shakapacker/* and instead use shakapacker.yml to load different configurations using RAILS_ENV.

The updated webpack.config.js sets `NODE_ENV` based on `RAILS_ENV`:

```javascript
const { generateWebpackConfig, merge } = require('shakapacker')
const webpack = require('webpack')

if (['development', 'test', ].includes(process.env.NODE_ENV)) {
  process.env.NODE_ENV = 'development'
}
else {
  process.env.NODE_ENV = 'production'
}

const webpackConfig = generateWebpackConfig()
const customConfig = {
  devServer: {
    watchFiles: {
      options: {
        ignored: '**/.*'
      }
    }
  },
  plugins: [
    new webpack.ProvidePlugin({
      $: 'jquery',
      jQuery: 'jquery',
    }),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV),
    })
  ]
}

module.exports = merge(webpackConfig, customConfig)
``` 

Still, I'd be interested to hear if I am doing this correctly.

**Comment by Judahmeek** [comment, marked as answer]
@jpwilksch I'm glad that you found a solution that works for you.

The relevant part of our documentation from the section you linked to is:

> Please note, `NODE_ENV` can either be set to production, development or test. This means you don't need to create additional environment files inside config/shakapacker/* and instead use shakapacker.yml to load different configurations using `RAILS_ENV`.

To clarify, Shakapacker supports custom `RAILS_ENV` values, but Shakapacker does not support custom `NODE_ENV` values because webpack does not support custom `NODE_ENV` values.

**Reply by ahangarha** [reply]
@Judahmeek Please share the docs with this text. It needs a minor update.

**Reply by jpwilksch** [reply]
@Judahmeek Thanks for your response. 

I understand that webpack does not support custom `NODE_ENV` values. However, I'd expect anyone using custom `RAILS_ENV` values is going to want these to map to a valid `NODE_ENV` value to avoid the errors I encountered. 

It would be beneficial if Shakapacker could provide a mechanism to do this mapping - either automatically or via `config/shakapacker.yml`
